### PR TITLE
339 semaphore tests

### PR DESCRIPTION
**Pull request by edvmorango**
#339

### CODE REVIEW

File: core/jvm/src/test/scala/scalaz/zio/SemaphoreSpec.scala [hunk 1]
```diff
@@ -53,6 +57,24 @@ class SemaphoreSpec extends AbstractRTSSpec {
     } yield () must_=== (()))
   }
 
+  def e6 = {
+    val n = 1L
+    unsafeRun(for {
+      s <- Semaphore(n)
+      _ <- s.acquireN(2).timeoutFail(new Exception("timeout"))(1.milli).attempt
+      permits <- IO.sleep(10.millis) *> s.available
+    } yield permits) must_=== 0
```
[the remark this answers is not in the document]
Comment (edvmorango) [reply core/jvm/src/test/scala/scalaz/zio/SemaphoreSpec.scala]: Hey @regiskuckaertz, thank you for the feedback, I had misunderstood the idea. I will work on the requested changes.

[the remark this answers is not in the document]
Comment (regiskuckaertz) [reply core/jvm/src/test/scala/scalaz/zio/SemaphoreSpec.scala]: You can literally copy-paste their tests, we should get to the exact same result.

**Comment by regiskuckaertz** [comment]
Try with this:
```scala
  final def withPermit[E, A](task: IO[E, A]): IO[E, A] =
    IO.bracket0[E, AcquireTasks, A](prepare(1L))(cleanup) { case (acquire, _) => acquire *> task }

  final def acquireN(requested: Long): IO[Nothing, Unit] =
    assertNonNegative(requested) *> IO.bracket0[Nothing, AcquireTasks, Unit](prepare(requested))(cleanup)(_._1)

  final private def prepare(n: Long): IO[Nothing, AcquireTasks] = {
    def restore(p: Promise[Nothing, Unit], n: Long): IO[Nothing, Unit] =
      state.modify {
        case Left(q) =>
          q.find(_._1 == p).fold(releaseN(n) -> Left(q))(x => releaseN(n - x._2) -> Left(q.filter(_._1 != p)))
        case Right(m) => IO.unit -> Right(m + n)
      }.void

    if (n == 0)
      IO.now((IO.unit, IO.unit))
    else
      Promise.make[Nothing, Unit].flatMap { p =>
        state.modify {
          case Right(m) if m >= n => (IO.unit, releaseN(n)) -> Right(m - n)
          case Right(m)           => (p.get, restore(p, n)) -> Left(IQueue(p -> (n - m)))
          case Left(q)            => (p.get, restore(p, n)) -> Left(q.enqueue(p -> n))
        }
      }
  }

  final private def cleanup[E, A](ops: AcquireTasks, res: ExitResult[E, A]): IO[Nothing, Unit] =
    res match {
      case ExitResult.Terminated(_) => ops._2
      case _                        => IO.unit
    }
```
where `type AcquireTasks = (IO[Nothing, Unit], IO[Nothing, Unit])`

**Comment by edvmorango** [comment]
@regiskuckaertz I followed what you said, I think it's done now.

File: core/shared/src/main/scala/scalaz/zio/Semaphore.scala [hunk 2]
```diff
@@ -10,6 +10,8 @@ import scala.collection.immutable.{ Queue => IQueue }
 
 final class Semaphore private (private val state: Ref[State]) extends Serializable {
 
+  type AcquireTasks = (IO[Nothing, Unit], IO[Nothing, Unit])
```
Comment (regiskuckaertz) [inline core/shared/src/main/scala/scalaz/zio/Semaphore.scala]: Let's move that into the companion.

File: core/shared/src/main/scala/scalaz/zio/Semaphore.scala [hunk 3]
```diff
@@ -22,26 +24,37 @@ final class Semaphore private (private val state: Ref[State]) extends Serializab
...
-      case (p, Right(n)) if n >= requested => p.complete(()) -> Right(n - requested)
-      case (p, Right(n))                   => IO.now(false)  -> Left(IQueue(p -> (requested - n)))
-      case (p, Left(q))                    => IO.now(false)  -> Left(q.enqueue(p -> requested))
-    }
+    IO.bracket0[E, AcquireTasks, A](prepare(1L))(cleanup) { case (acquire, _) => acquire *> task }
+
+  final def acquireN(requested: Long): IO[Nothing, Unit] =
+    assertNonNegative(requested) *> IO.bracket0[Nothing, AcquireTasks, Unit](prepare(requested))(cleanup)(_._1)
+
+  final private def prepare(n: Long): IO[Nothing, AcquireTasks] = {
```
Comment (regiskuckaertz) [inline core/shared/src/main/scala/scalaz/zio/Semaphore.scala]: All credit should go to @mpilquist for finding out the solution in https://github.com/typelevel/cats-effect/pull/403, so let's mention that in a comment.

File: core/shared/src/main/scala/scalaz/zio/Semaphore.scala [hunk 4]
```diff
@@ -22,26 +24,37 @@ final class Semaphore private (private val state: Ref[State]) extends Serializab
...
   final def withPermit[E, A](task: IO[E, A]): IO[E, A] =
-    IO.bracket[E, Unit, A](acquire)(_ => release)(_ => task)
-
-  final def acquireN(requested: Long): IO[Nothing, Unit] = {
-    val acquire: (Promise[Nothing, Unit], State) => (IO[Nothing, Boolean], State) = {
-      case (p, Right(n)) if n >= requested => p.complete(()) -> Right(n - requested)
-      case (p, Right(n))                   => IO.now(false)  -> Left(IQueue(p -> (requested - n)))
-      case (p, Left(q))                    => IO.now(false)  -> Left(q.enqueue(p -> requested))
-    }
+    IO.bracket0[E, AcquireTasks, A](prepare(1L))(cleanup) { case (acquire, _) => acquire *> task }
```
Comment (regiskuckaertz) [inline core/shared/src/main/scala/scalaz/zio/Semaphore.scala]: `_._1 *> task` is more concise.

File: core/shared/src/main/scala/scalaz/zio/Semaphore.scala [hunk 5]
```diff
@@ -22,26 +24,37 @@ final class Semaphore private (private val state: Ref[State]) extends Serializab
...
+
+  final def acquireN(requested: Long): IO[Nothing, Unit] =
```
Comment (regiskuckaertz) [inline core/shared/src/main/scala/scalaz/zio/Semaphore.scala]: Just saw that now we use `n` everywhere so let’s stick to that

File: core/shared/src/main/scala/scalaz/zio/Semaphore.scala (hunk 2 above)
Comment (edvmorango) [reply core/shared/src/main/scala/scalaz/zio/Semaphore.scala]: @regiskuckaertz It seemed right to move AcquireTaks to internals, did I do it right?

Comment (regiskuckaertz) [reply core/shared/src/main/scala/scalaz/zio/Semaphore.scala]: Yep

**Comment by ghostdogpr** [comment]
> Thanks for your great work on this @ghostdogpr 

I don't deserve any credit for this ticket 😅but @edvmorango does!

**Comment by edvmorango** [comment]
Thank you guys, for the help and feedback.

**Pull request merged**